airbyte-commons-protocol/src/main/java/io/airbyte/commons/protocol/migrations/util/RecordMigrations.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.commons.protocol.migrations.util;
+
+import static io.airbyte.protocol.models.JsonSchemaReferenceTypes.REF_KEY;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.validation.json.JsonSchemaValidator;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public class RecordMigrations {
```

Comment:
fyi @akashkulk - this + SchemaMigrations is mostly useful for future protocol updates where we add a new data type. Those migrations should be a lot easier to write - example for if we add a jsonb type in protocol v6: (there's still some copypasta in here, but meh whatever)
```java
private static boolean isJsonbSchema(JsonNode schema) {
  return schema.hasNonNull("$ref") && schema.get("$ref").asText().equals("WellKnownTypes.json#/definitions/NewJsonbType");
}

@Override
public io.airbyte.protocol.models.v0.AirbyteMessage downgrade(AirbyteMessage oldMessage, ConfiguredAirbyteCatalog catalog) {
  io.airbyte.protocol.models.v0.AirbyteMessage newMessage = Jsons.object(
      Jsons.jsonNode(oldMessage),
      io.airbyte.protocol.models.v0.AirbyteMessage.class);
  if (oldMessage.getType() == Type.CATALOG) {
    for (io.airbyte.protocol.models.v0.AirbyteStream stream : newMessage.getCatalog().getStreams()) {
      JsonNode schema = stream.getJsonSchema();
      SchemaMigrations.mutateSchemas(
          HypotheticalV6Migration::isJsonbSchema,
          s -> {
            ObjectNode o = (ObjectNode) s;
            o.removeAll();
            o.put("$ref", "WellKnownTypes.json#/definitions/String")
          },
          schema);
    }
  } else if (oldMessage.getType() == Type.RECORD) {
    io.airbyte.protocol.models.v0.AirbyteRecordMessage record = newMessage.getRecord();
    Optional<ConfiguredAirbyteStream> maybeStream = catalog.getStreams().stream()
        .filter(stream -> Objects.equals(stream.getStream().getName(), record.getStream())
            && Objects.equals(stream.getStream().getNamespace(), record.getNamespace()))
        .findFirst();
    if (maybeStream.isPresent()) {
      JsonNode schema = maybeStream.get().getStream().getJsonSchema();
      JsonNode oldData = record.getData();
      MigratedNode downgradedNode = RecordMigrations.mutateDataNode(
          validator,
          HypotheticalV6Migration::isJsonbSchema,
          // V5 doesn't have a JSONB type, so just stringify this thing
          // JSONB is, by definition, any JSON value, so we always match the schema.
          (s, d) -> new MigratedNode(new TextNode(Jsons.serialize(d)), true),
          oldData,
          schema);
      record.setData(downgradedNode.node());
    }
  }
  return newMessage;
}

@Override
public AirbyteMessage upgrade(io.airbyte.protocol.models.v0.AirbyteMessage oldMessage, ConfiguredAirbyteCatalog catalog) {
  // Noop - since we're just adding a new data type, all old messages are valid as new messages
  return Jsons.object(
      Jsons.jsonNode(oldMessage),
      AirbyteMessage.class);
}
```